Sources/NIOHPACK/SimpleRingBuffer.swift

```diff
@@ -0,0 +1,933 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2018 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+public enum RingBufferError {
+    /// Error type thrown when a write would overrun the bounds of a
+    /// CircularByteBuffer.
+    public struct BufferOverrun : Error, Equatable {
+        /// The amount by which the write would overrun.
+        public let amount: Int
+        
+        public init(amount: Int) {
+            self.amount = amount
+        }
+    }
+}
+
+@usableFromInline
+internal func _toIndex(_ value: Int) -> SimpleRingBuffer.Index {
+    return SimpleRingBuffer.Index(truncatingIfNeeded: value)
+}
+
+@usableFromInline
+internal func _toCapacity(_ value: Int) -> SimpleRingBuffer.Capacity {
+    return SimpleRingBuffer.Capacity(truncatingIfNeeded: value)
+}
+
+@usableFromInline
+func _wrappingDifference(from: SimpleRingBuffer.Index, to: SimpleRingBuffer.Index, within: SimpleRingBuffer.Capacity) -> SimpleRingBuffer.Capacity {
+    if from <= to {
+        return to - from
+    } else {
+        return (within - from) + to
+    }
+}
+
+public struct SimpleRingBuffer {
```

Comment:
Is there any reason this shouldn't just wrap a `ByteBuffer`, rather than bring in all the implementation details of `ByteBuffer`?

Reply:
I tried a few different things, but from outside NIO itself I found it quite awkward to manage the internals of the `ByteBuffer` directly. I found I was doing basically all the same calculations, and the only real difference was that I was either using the `ByteBuffer` API or the `UnsafeMutablePointer` API for the underlying work. Since this had only a single client and one use-case I cut it down to just mirroring the `String` handling from `ByteBuffer` and none of the fancy raw pointer stuff, with the exception of `withVeryUnsafePointer` which I'd used somewhere in a private method somewhere else, or so I seemed to recall.

There's definitely an argument for something like `ByteBuffer` that actually operates as a fixed-capacity circular collection, but the right place for that would be inside NIO proper, sharing some of the implementation details with `ByteBuffer` directly. I could probably put one together as a PR for swift-nio, but that would happen later; after that I'd change the implementation of the header tables here to use that type instead.

Reply:
I guess I'm a bit uncertain about why maintaining two indices outside of ByteBuffer and using the set/get APIs was more complex than bringing this code in. What am I missing?

Reply:
I'll have another go; maybe it'll become simpler now that I've really pared down the ring buffer's API.

Reply:
Okay, that's done, and I've pruned a lot of the speculative API that wasn't being used yet.